src/Wrapper/Csv.php

```diff
@@ -149,16 +149,41 @@ public function getResults()
     /**
      * Write item data to file
      *
-     * @param array $data
+     * @param Product $product
      * @return $this
+     * @throws CriticalException
      */
-     public function wrapItem(array $data)
+     public function wrapItem(Product $product)
      {
          // Check if headers already set
          if (!$this->headersSetFlag) {
-             $this->setHeaders($data);
+             $this->setHeaders($product->getFieldNames());
          }
 
+         $data['id'] = $product->getFieldId();
+         $data['ordernumber'] = $product->getFieldOrdernumber();
+         $data['name'] = $product->getFieldName();
+         $data['summary'] = $product->getFieldSummary();
+         $data['description'] = $product->getFieldDescription();
+         $data['price'] = $product->getFieldPrice();
+         $data['instead'] = $product->getFieldInstead();
+         $data['maxprice'] = $product->getFieldMaxprice();
+         $data['taxrate'] = $product->getFieldTaxrate();
+         $data['url'] = $product->getFieldUrl();
+         $data['image'] = $product->getFieldImage();
+         $data['base_unit'] = $product->getFieldBaseUnit();
+         $data['package_size'] = $product->getFieldPackageSize();
+         $data['price_id'] = $product->getFieldPriceId();
+         $data['attributes'] = $product->getFieldAttributes();
+         $data['keywords'] = $product->getFieldKeywords();
+         $data['groups'] = $product->getFieldGroups();
+         $data['bonus'] = $product->getFieldBonus();
+         $data['sales_frequency'] = $product->getFieldSalesFrequency();
+         $data['date_added'] = $product->getFieldDateAdded();
+         $data['sort'] = $product->getFieldSort();
+         $data['main_variation_id'] = $product->getFieldMainVariationId();
+         $data['variation_id'] = $product->getFieldVariationId();
+
          $data = $this->convertData($data);
 
          fputcsv($this->getStream(), $data, self::CSV_DELIMITER, self::CSV_ENCLOSURE);
```

Comment:
Do you know is it possible to write part of line to the csv file? Because right now it still saves all $data array in memory.

Reply:
Check out method `serializeItems` https://github.com/findologic/libflexport/blob/develop/src/FINDOLOGIC/Export/CSV/CSVExporter.php. It goes through all of them.

Be careful, moment you do this, you have pretty much gained nothing from the entire streaming approach:

```php
         $data['ordernumber'] = $product->getFieldOrdernumber();
         $data['name'] = $product->getFieldName();
         $data['summary'] = $product->getFieldSummary();
         $data['description'] = $product->getFieldDescription();
         $data['price'] = $product->getFieldPrice();
         $data['instead'] = $product->getFieldInstead();
         $data['maxprice'] = $product->getFieldMaxprice();
         $data['taxrate'] = $product->getFieldTaxrate();
         $data['url'] = $product->getFieldUrl();
         $data['image'] = $product->getFieldImage();
         $data['base_unit'] = $product->getFieldBaseUnit();
         $data['package_size'] = $product->getFieldPackageSize();
         $data['price_id'] = $product->getFieldPriceId();
         $data['attributes'] = $product->getFieldAttributes();
         $data['keywords'] = $product->getFieldKeywords();
         $data['groups'] = $product->getFieldGroups();
         $data['bonus'] = $product->getFieldBonus();
         $data['sales_frequency'] = $product->getFieldSalesFrequency();
         $data['date_added'] = $product->getFieldDateAdded();
         $data['sort'] = $product->getFieldSort();
         $data['main_variation_id'] = $product->getFieldMainVariationId();
         $data['variation_id'] = $product->getFieldVariationId();
```

You need to do it one by one:

```php
$availableData = ['id', 'ordernumber', 'and so on...'];
foreach ($availableData as $value) {
    // Create a class that converts this. Something like that is pretty common.
    $method = 'get' . $this->snakeCaseToCamelCaseConverter->convert($value);
    // Write data each line:
    $this->writeToCsv($value, $product->$method());
}
```
Otherwise you are writing everything into memory, which we want to get rid of.

There are also some fields which are converted to the correct format later in `$data = $this->convertData($data)`. The data should be pre-formatted, before it gets to this method. This way the wrapper knows to which row the data should be written.